src/java.base/share/classes/jdk/internal/reflect/ReflectionFactory.java

```diff
@@ -623,74 +587,135 @@ public final boolean hasStaticInitializerForSerialization(Class<?> cl) {
-        String val = props.getProperty("sun.reflect.noInflation");
-        if (val != null && val.equals("true")) {
-            noInflation = true;
+        Config c = config;
+        if (c == null) {
+            config = c = new Config(true);
         }
+        return c;
+    }
 
-        val = props.getProperty("sun.reflect.inflationThreshold");
-        if (val != null) {
-            try {
-                inflationThreshold = Integer.parseInt(val);
-            } catch (NumberFormatException e) {
-                throw new RuntimeException("Unable to parse property sun.reflect.inflationThreshold", e);
-            }
-        }
-        val = props.getProperty("jdk.reflect.useDirectMethodHandle");
-        if (val != null) {
-            if (val.equals("false")) {
-                useDirectMethodHandle = 0;
-            } else if (val.equals("methods")) {
-                useDirectMethodHandle = METHOD_MH_ACCESSOR;
-            } else if (val.equals("fields")) {
-                useDirectMethodHandle = FIELD_MH_ACCESSOR;
-            }
-        }
-        val = props.getProperty("jdk.reflect.useNativeAccessorOnly");
-        if (val != null && val.equals("true")) {
-            useNativeAccessorOnly = true;
-        }
+    /**
+     * The configurations exist as an object to avoid race conditions.
+     * See bug 8261407.
+     */
+    private static final class Config {
+        private static final Config fallback = new Config(false);
+
+        //
+        // "Inflation" mechanism. Loading bytecodes to implement
+        // Method.invoke() and Constructor.newInstance() currently costs
+        // 3-4x more than an invocation via native code for the first
+        // invocation (though subsequent invocations have been benchmarked
+        // to be over 20x faster). Unfortunately this cost increases
+        // startup time for certain applications that use reflection
+        // intensively (but only once per class) to bootstrap themselves.
+        // To avoid this penalty we reuse the existing JVM entry points
+        // for the first few invocations of Methods and Constructors and
+        // then switch to the bytecode-based implementations.
+        //
+        private final boolean noInflation;
+        private final int     inflationThreshold;
+
+        //
+        // New implementation uses direct invocation of method handles
+        private static final int METHOD_MH_ACCESSOR      = 0x1;
+        private static final int FIELD_MH_ACCESSOR       = 0x2;
+        private static final int ALL_MH_ACCESSORS        = METHOD_MH_ACCESSOR|FIELD_MH_ACCESSOR;
+
+        private final int     useDirectMethodHandle;
+        private final boolean useNativeAccessorOnly;  // for testing only
+
+        // true if deserialization constructor checking is disabled
+        private final boolean disableSerialConstructorChecks;
+
+        private Config(boolean getProperties) {
```

Comment:
I suggest to make this a static `Config.getInstance()` method to return a `Config` instance and that should assert if this is called after the module system is initialized.   The `Config` constructor is a simple constructor taking one parameter for each field and just set all fields to the parameter value.

Reply:
Can I just write the config class as a record, or does it generate too much boilerplate?

Reply:
Worth a try.   Even the regular class, the constructor taking 5 fields isn't too bad to me.  In a near future, I hope to remove the old core reflection implementation, `noInflation` and `inflationThreshold` will be removed and fewer fields.

Reply:
I made a commit with the config class converted into a record. Apparently the tests are passing, and I would assume it would be feasible. Should I apply it?
https://github.com/liachmodded/jdk/commit/8cf5af417a6f906e9fc0c878d60731d6f026b528

Reply:
yes and I can take a closer it.

Indy is ready to use very early during VM initialization before initPhase2 where the module system is initialized.   AFAIU, indy is needed for the object methods for records i.e. `equals`, `hashCode`, and `toString`.  Just record object instantiation and accessing its final fields don't use indy.